### PR TITLE
source-rarible contribution from bishalbera

### LINKED ISSUE #46035: [connector-request] Rarible
This is a **suggestion** to create a connector for **Rarible** using the UI Connector Builder.
The website for the connector is https://rarible.com.

If you want to try, claim this issue and start working on it following the steps below.

### Steps:
1. Comment in the issue and wait to be assigned to start working on it.
2. Map the API endpoints (get approval to move to 3 step)
3. Describe steps to get credentials.
4. Create the connector using the UI Builder

> [!IMPORTANT]
> Airbyte doesn't have sandbox credentials for this connector. You must have access/credentials to the service provider to create the connector.

> [!CAUTION]
> This is a **suggestion** and there may be cases where creating the connector using the Builder won't be possible. To minimize the risk of investing a lot of time directly in creating the connector, we strongly recommend to follow steps above.

### PR DESCRIPTION
**Pull request by bishalbera**
## What

This PR adds a new connector for source Rarible (source-rarible).

Contributor-provided connector description:

This connector integrates Rarible&#39;s API with Airbyte, enabling seamless data synchronization of NFT activities, collections, and ownership across multiple blockchains. It supports fetching  ownership, orders and other essential details to facilitate real-time updates and analytics for NFT marketplaces.
## Details

<details><summary>Expand for list of spec properties and streams</summary>

### Configuration

| Input | Type | Description | Default Value |
|-------|------|-------------|---------------|
| `api_key` | `string` | API Key.  |  |
| `collection_id` | `string` | Collection Id.  |  |
| `blockchain` | `string` | Blockchain.  |  |
| `owner` | `string` | Owner.  |  |
| `maker` | `string` | Maker. ETHEREUM OR POLYGON or leave empty to get orders from all blockchains |  |
| `start_date` | `string` | Start date.  |  |
| `lastUpdatedFrom` | `integer` | lastUpdatedFrom.  |  |
| `lastUpdatedTo` | `number` | lastUpdatedTo.  |  |
| `currency` | `string` | Currency.  | USD |

### Streams
| Stream Name | Primary Key | Pagination | Supports Full Sync | Supports Incremental |
|-------------|-------------|------------|---------------------|----------------------|
| nfts |  | DefaultPaginator | ✅ |  ✅  |
| collections | id | No pagination | ✅ |  ❌  |
| collection_stats |  | No pagination | ✅ |  ❌  |
| orders | id | DefaultPaginator | ✅ |  ❌  |
| activities | id | DefaultPaginator | ✅ |  ❌  |
| user_owned_nfts | id | DefaultPaginator | ✅ |  ❌  |
| users_bid_orders | id | DefaultPaginator | ✅ |  ❌  |

</details>

## Reviewer checklist
- [ ] Run `/format-fix` to resolve any formatting errors
- [ ] Help the contributor upload an icon that follows our icon design guidelines
- [ ] Click into the CI workflows that wait for a maintainer to run them, which should trigger CI runs

### CODE REVIEW

File: airbyte-integrations/connectors/source-rarible/manifest.yaml [hunk 1]
```diff
@@ -0,0 +1,1739 @@
...
+      type: DeclarativeStream
+      name: activities
+      retriever:
+        type: SimpleRetriever
+        requester:
+          $ref: "#/definitions/base_requester"
+          path: /v0.1/activities/sync
+          http_method: GET
+          request_parameters:
+            blockchain: "{{ config['blockchain'] }}"
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-rarible/manifest.yaml:106]: Why not use the all endpoint? https://docs.rarible.org/reference/getallactivities

File: airbyte-integrations/connectors/source-rarible/manifest.yaml [hunk 2]
```diff
@@ -0,0 +1,1739 @@
...
+        schema:
+          $ref: "#/schemas/collections"
+    statistics:
+      type: DeclarativeStream
+      name: statistics
+      retriever:
+        type: SimpleRetriever
+        requester:
+          $ref: "#/definitions/base_requester"
+          path: v0.1/data/collections/{{ config['collection_id'] }}/stats
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-rarible/manifest.yaml:67]: Why not use collection stream as parent of this one?

**Review by marcosmarxm — CHANGES_REQUESTED** [review]
Something strange is forcing the user to provide `onwer`, `maker`. and `collection_id` to use the connector. Maybe transform these to optional.

File: airbyte-integrations/connectors/source-rarible/manifest.yaml (hunk 1 above)
Comment (bishalbera) [reply airbyte-integrations/connectors/source-rarible/manifest.yaml:106]: @marcosmarxm it says its deprecated 
Screenshot 2024-10-09 at 8 21 00 AM

File: airbyte-integrations/connectors/source-rarible/manifest.yaml (hunk 2 above)
Comment (bishalbera) [reply airbyte-integrations/connectors/source-rarible/manifest.yaml:67]: yes I'll update it . At that time I didnt knew about the parent stream thing

**Comment by bishalbera** [comment]
> Something strange is forcing the user to provide `onwer`, `maker`. and `collection_id` to use the connector. Maybe transform these to optional.

These are required params for their respected streams. We cant make those optional

File: airbyte-integrations/connectors/source-rarible/manifest.yaml (hunk 1 above)
Comment (marcosmarxm) [reply airbyte-integrations/connectors/source-rarible/manifest.yaml:119]: oh! thanks!

File: airbyte-integrations/connectors/source-rarible/manifest.yaml (hunk 2 above)
Comment (marcosmarxm) [reply airbyte-integrations/connectors/source-rarible/manifest.yaml]: 👍

**Review by marcosmarxm — CHANGES_REQUESTED** [review]
`nfts` stream: it is missing pagination, [check documentation](https://docs.rarible.org/reference/getallitems) it has the parameter `continuation`, `size` to retrive more data instead of default 50 and `lastUpdatedFrom` and `lastUpdatedTo` which allows to implement incremental.

- [x] implement pagination for nft stream
- [x] implement incremental for nft stream
- [x] increase page size to max allow by the API

`collections` the same but doesn't allow incremental.
- [ ] implement pagination 
- [ ] increase page size to max allow by the API

`collection_stats`: probably should be a good idea implement a enum to select what currency the user want to retrieve the data. This isn't working using the Parent stream anymore. Also you must implement a transformation to include the `collection_id` in the record response to make data readable.
- [ ] fix inject path using collection stream partition
- [ ] add enum parameter in config to select what currency user want to use

`orders` missing pagination and increase size
- [ ] fix pagination
- [ ] fix max page size
- [ ] make sure items are ordered `updated_at` asc

I didn't look at other streams yet but this gives you some points to continue.

About pagination the API returns a key-value `continuation`.
````
{
  "continuation": "1729095982264_BASE:0x394d6f3fbf1849ceae548752bd0428a2b847a33932bfa14101479ad35c93f522",
  "orders": [
    {
```

**Comment by bishalbera** [comment]
> `nfts` stream: it is missing pagination, [check documentation](https://docs.rarible.org/reference/getallitems) it has the parameter `continuation`, `size` to retrive more data instead of default 50 and `lastUpdatedFrom` and `lastUpdatedTo` which allows to implement incremental.
> 
> * [ ]  implement pagination for nft stream
> * [ ]  implement incremental for nft stream
> * [ ]  increase page size to max allow by the API
> 
> `collections` the same but doesn't allow incremental.
> 
> * [ ]  implement pagination
> * [ ]  increase page size to max allow by the API
> 
> `collection_stats`: probably should be a good idea implement a enum to select what currency the user want to retrieve the data. This isn't working using the Parent stream anymore. Also you must implement a transformation to include the `collection_id` in the record response to make data readable.
> 
> * [ ]  fix inject path using collection stream partition
> * [ ]  add enum parameter in config to select what currency user want to use
> 
> `orders` missing pagination and increase size
> 
> * [ ]  fix pagination
> * [ ]  fix max page size
> * [ ]  make sure items are ordered `updated_at` asc
> 
> I didn't look at other streams yet but this gives you some points to continue.
> 
> About pagination the API returns a key-value `continuation`.
> 
> ```
> {
>   "continuation": "1729095982264_BASE:0x394d6f3fbf1849ceae548752bd0428a2b847a33932bfa14101479ad35c93f522",
>   "orders": [
>     {
> ```
> ```

@marcosmarxm `lastUpdatedFrom`  and `lastUpdatedTo` parameter is a type of int64(timestamp in miliseconds) but if I want to use it in incremental sync it requires to be datetime which is giving me `400 bad status code` error. However I can use these parameters as user input and set type to interger and let the user enter that value. Also I didnt understand how should i use the `size` parameter.

**Review by marcosmarxm — COMMENTED** [review]
It looks there is a bug in the connector builder @bishalbera. For that reason ask you some patience until I get answers from the connector team.

The 400 error you're getting, I got the same, but after recreate the stream it was gone. The problem is returned. 

New Estimated Next Review: 2024-10-25

**Comment by marcosmarxm** [comment]
Hello 👋 Due to the numerous contributions happening, I decided to create this [view](https://github.com/orgs/airbytehq/projects/104/views/3) to help you know when I plan to review your contribution. If your connector has some limited-time credential feel free to reach me out.

**Review by marcosmarxm — APPROVED** [review]
I don't think team will publish changes needed to fix the problem to this connector. But I extensily tested and there is only one limitation in the builder. After this get fixed I'll merge this contribution.

**Comment by marcosmarxm** [comment]
Closing this to clean backlog and revamp builder process. I'm going to submit a new one in the future after get the bug fix.

**Pull request closed without merging**